node/node.go

```diff
@@ -0,0 +1,314 @@
+			return ErrDatadirUsed
+		}
+		return err
+	}
+	// Start each of the services
+	started := []string{}
+	for id, service := range services {
+		// Start the next service, stopping all previous upon failure
+		if err := service.Start(running); err != nil {
+			for _, id := range started {
+				services[id].Stop()
+			}
+			running.Stop()
+
+			return err
+		}
+		// Mark the service started for potential cleanup
+		started = append(started, id)
+	}
+	// Finish initializing the startup
+	n.services = services
+	n.server = running
+	n.stop = make(chan struct{})
+
+	return nil
+}
+
+// Stop terminates a running node along with all it's services. In the node was
+// not started, an error is returned.
+func (n *Node) Stop() error {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	// Short circuit if the node's not running
+	if n.server == nil {
+		return ErrNodeStopped
+	}
+	// Otherwise terminate all the services and the P2P server too
+	failure := &StopError{
+		Services: make(map[string]error),
+	}
+	for id, service := range n.services {
+		if err := service.Stop(); err != nil {
+			failure.Services[id] = err
+		}
+	}
+	n.server.Stop()
+
+	n.services = nil
+	n.server = nil
+	close(n.stop)
+
+	if len(failure.Services) > 0 {
+		return failure
+	}
+	return nil
+}
+
+// Wait blocks the thread until the node is stopped. If the node is not running
+// at the time of invocation, the method immediately returns.
+func (n *Node) Wait() {
+	n.lock.RLock()
+	if n.server == nil {
+		return
+	}
+	stop := n.stop
+	n.lock.RUnlock()
+
+	<-stop
+}
+
+// Restart terminates a running node and boots up a new one in its place. If the
+// node isn't running, an error is returned.
+func (n *Node) Restart() error {
+	if err := n.Stop(); err != nil {
+		return err
+	}
+	if err := n.Start(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Server retrieves the currently running P2P network layer. This method is meant
+// only to inspect fields of the currently running server, life cycle management
+// should be left to this Node entity.
+func (n *Node) Server() *p2p.Server {
+	n.lock.RLock()
+	defer n.lock.RUnlock()
+
+	return n.server
+}
+
+// Service retrieves a currently running service registered under a given id.
+func (n *Node) Service(id string) Service {
+	n.lock.RLock()
+	defer n.lock.RUnlock()
+
+	// Short circuit if the node's not running
+	if n.server == nil {
+		return nil
+	}
+	return n.services[id]
+}
+
+// SingletonService retrieves a currently running service using a specific type
+// implementing the Service interface. This is a utility function for scenarios
+// where it is known that only one instance of a given service type is running,
+// allowing to access services without needing to know their specific id with
+// which they were registered. Note, this method uses reflection, so do not run
+// in a tight loop.
+func (n *Node) SingletonService(service interface{}) (string, error) {
```

Comment:
Originally I also had only the id based version. The problem that occurred was that a lot of code needs to be able to retrieve an ethereum object from the stack (xeth, rpc to name two monsters), so you inherently hard code the name of the object into these. Therefore you can't really initialize a custom stack, because the ids are hard coded all over the place.

(I had quite a lot of issues exactly because of this because the live code used one id's and my tests used another; which was then a pain in the ass to rewrite because I had no clue where I'm referencing the old id and where the new)

My suggestion would be to keep this in for now, and lets see how far we can remove unneeded references after refactoring the RPC and removing the Xeth layer. Though if you oppose it, I can revert to the base version.